src/tools/pkcs11-tool.c

```diff
@@ -2510,7 +2516,15 @@ static void sign_data(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
 			sz = read(fd, in_buffer, sizeof(in_buffer));
 		} while (sz > 0);
 
-		sig_len = sizeof(sig_buffer);
+		/* Signature buffer may have been allocated above */
+		free(sig_buffer);
+		sig_buffer = NULL;
+		rv = p11->C_SignFinal(session, sig_buffer, &sig_len);
+		if (rv != CKR_OK)
+			p11_fatal("C_SignFinal", rv);
+		sig_buffer = malloc(sig_len);
```

Comment:
nit: Do we need to free and allocate the `sig_buffer` here? Given that this is the same operation with the same key, I would expect the signature size would be same for both `C_Sign` and `C_SignFinal` operation. 
But this is really just a corner case. Making sure the length of the old buffer is right would complicate the code probably eve more.